pkg/ccl/backupccl/backup_test.go

```diff
@@ -10125,3 +10144,192 @@ func TestBackupRestoreSeparateExplicitIsDefault(t *testing.T) {
+	for i := 0; i < numCheckpoints; i++ {
+		checkpoints = append(checkpoints, newTimestampedCheckpointFileName())
+		// Occasionally, we call newTimestampedCheckpointFileName() in succession
+		// too fast and the timestamp is the same. So wait for a moment to
+		// avoid that.
+		time.Sleep(time.Millisecond)
+	}
+
+	const numAccounts = 1000
+	tc, _, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
+	defer cleanupFn()
+	execCfg := tc.Server(0).ExecutorConfig().(sql.ExecutorConfig)
+	ctx := context.Background()
+
+	for _, tc := range []struct {
+		name string
+	}{
+		{
+			"azure",
+		},
+		{
+			"fileTable",
+		},
+		{
+			"gcs",
+		},
+		{
+			"localFile",
+		},
+		{
+			"s3",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			var uri string
+			switch tc.name {
+			case "s3":
+				// If environment credentials are not present, we want to
+				// skip the S3 test.
+				accessKeyID := os.Getenv("AWS_ACCESS_KEY_ID")
+				if accessKeyID == "" {
+					skip.IgnoreLint(t, "AWS_ACCESS_KEY_ID env var must be set")
+				}
+				secretAccessKey := os.Getenv("AWS_SECRET_ACCESS_KEY")
+				if secretAccessKey == "" {
+					skip.IgnoreLint(t, "AWS_SECRET_ACCESS_KEY env var must be set")
+				}
+				bucket := os.Getenv("AWS_S3_BUCKET")
+				if bucket == "" {
+					skip.IgnoreLint(t, "AWS_S3_BUCKET env var must be set")
+				}
+
+				uri = amazon.S3URI(bucket, "backup-test-lexicographical",
+					&roachpb.ExternalStorage_S3{AccessKey: accessKeyID, Secret: secretAccessKey, Region: "us-east-1"},
+				)
+			case "azure":
+				account := os.Getenv("AZURE_ACCOUNT_NAME")
+				key := os.Getenv("AZURE_ACCOUNT_KEY")
+				bucket := os.Getenv("AZURE_CONTAINER")
+
+				if account == "" || key == "" || bucket == "" {
+					skip.IgnoreLint(t, "AZURE_ACCOUNT_NAME, AZURE_ACCOUNT_KEY, AZURE_CONTAINER must all be set")
+				}
+
+				uri = fmt.Sprintf("azure://%s/%s?%s=%s&%s=%s",
+					bucket, "backup-test-lexicographical",
+					azure.AzureAccountKeyParam, url.QueryEscape(key),
+					azure.AzureAccountNameParam, url.QueryEscape(account))
+			case "gcs":
+				bucket := os.Getenv("GS_BUCKET")
+				if bucket == "" {
+					skip.IgnoreLint(t, "GS_BUCKET env var must be set")
+				}
+
+				credentials := os.Getenv("GS_JSONKEY")
+				if credentials == "" {
+					skip.IgnoreLint(t, "GS_JSONKEY env var must be set")
+				}
+				encoded := base64.StdEncoding.EncodeToString([]byte(credentials))
+				uri = fmt.Sprintf("gs://%s/%s?%s=%s",
+					bucket,
+					"backup-test-lexicographical",
+					gcp.CredentialsParam,
+					url.QueryEscape(encoded),
+				)
+			case "fileTable":
+				uri = "userfile:///a"
+			case "localFile":
+				uri = "nodelocal://0/a"
+			}
+			store, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, uri, security.RootUserName())
+			require.NoError(t, err)
+			for _, checkpoint := range checkpoints {
+				var desc []byte
+				require.NoError(t, writeNewCheckpointFile(ctx, execCfg.Settings, store, checkpoint, desc))
```

Comment:
one quick comment, can we rand.Shuffle the `checkpoints` before we write them? This way we're doubly sure the List with limit 1 is independent of the order of writing.